### PR TITLE
Remove SymbolId - Option1

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Previously SymbolId and SymbolKey served nearly the same purpose.  They were a general way of describing a symbol that wouldn't keep a compilation alive and could be used to resolve symbolic values across symbols.  SymbolId was added because SymbolKey was not serializable, and could not be stored as a string property in a completion item.  This change updates SymbolKey so that it can now be serialized to/from Json.  With that change in place SymbolId was removable.

As part of this, i needed to expand SymbolKey a bit.  Previously, SymbolKey was just used for symbols for features like FAR.  In those cases, FAR always got a symbol key for the 'original definition' of a symbol.  For features like Completion though we need symbols for things like 'reduced methods'.  This required fleshing out some of the keys a bit.  But it was not a lot of work to do this.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
Fixes https://github.com/dotnet/roslyn/issues/11403

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
tagging @jasonmalinowski @rchande @jmarolf 

This now includes Ravi's SymbolId fix as well.

**Comment by CyrusNajmabadi** [comment]
Tagging @mattwar as well.

File: src/EditorFeatures/CSharpTest/SymbolId/SymbolKeyTestBase.cs [hunk 1]
```diff
@@ -77,8 +77,14 @@ internal static void ResolveAndVerifySymbol(ISymbol symbol1, ISymbol symbol2, Co
         internal static ISymbol ResolveSymbol(ISymbol originalSymbol, Compilation targetCompilation, SymbolKeyComparison comparison)
         {
             var sid = SymbolKey.Create(originalSymbol, CancellationToken.None);
-            var symInfo = sid.Resolve(targetCompilation, (comparison & SymbolKeyComparison.IgnoreAssemblyIds) == SymbolKeyComparison.IgnoreAssemblyIds);
 
+            // Verify that serialization works.
+            var serialized = sid.Serialize();
+            var deserialized = SymbolKey.Deserialize(serialized);
+            var comparer = SymbolKey.GetComparer(ignoreCase: false, ignoreAssemblyKeys: false);
+            Assert.True(comparer.Equals(sid, deserialized));
```
Comment (jmarolf) [inline src/EditorFeatures/CSharpTest/SymbolId/SymbolKeyTestBase.cs]: Could you add a failure message here so this is easier to diagnose?

File: src/EditorFeatures/VisualBasicTest/SymbolId/SymbolKeyTestBase.vb [hunk 2]
```diff
@@ -71,6 +71,14 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.SymbolId
 
         Friend Shared Function ResolveSymbol(originalSymbol As ISymbol, targetCompilation As Compilation, comparison As SymbolIdComparison) As ISymbol
             Dim sid = SymbolKey.Create(originalSymbol, CancellationToken.None)
+
+            ' Verify that serialization works.
+            Dim serialized = sid.Serialize()
+            Dim deserialized = SymbolKey.Deserialize(serialized)
+
+            Dim comparer = SymbolKey.GetComparer(ignoreCase:=False, ignoreAssemblyKeys:=False)
+            Assert.True(comparer.Equals(sid, deserialized))
```
Comment (jmarolf) [inline src/EditorFeatures/VisualBasicTest/SymbolId/SymbolKeyTestBase.vb]: Could you add a failure message here so this is easier to diagnose?

**Comment by mattwar** [comment]
This is a bad idea. It is way more important to have a small inexpensive string encoding for a symbol than we do to have the satisfaction that we've unified these pieces of code. Json serialization of SymbolKey is not sufficient. The whole process is going to take considerable more memory and generate a lot mor GC churn. In addition as we move to make more APIs public and available as protocols themselves the more we are going to want a simple publishable encoding of symbols. The only reason I didn't make SymbolId public already is to give us time to pick the best encoding.

**Comment by CyrusNajmabadi** [comment]
> This is a bad idea. It is way more important to have a small inexpensive string encoding

If this is actually an issue, we can switch to manually reading/writing Json with Json.net's extensibility points.

The point here is to not have two entirely separate, duplicated pieces of functionality in our code.  Better to have one, and to make it good enough for all purposes than to have two and have to maintain both.  This is especially important as the duplication of SymbolKey into SymbolId meant we had a crashing bug we had to fix twice.  

Going forward, it means we have to maintain both as well. For example, @jcouv added support for tuple types in SymbolKey.  However, there is no such support in SymbolId making the implementations diverge.

Right now Json encoding is an implementation detail, but one that allows us to properly unify on a single representation.  If we discover we need to optimize this bit we can now do so trivially, while still sticking with SymbolKey and not needing an entirely duplicated system.

> In addition as we move to make more APIs public and available as protocols themselves the more we are going to want a simple publishable encoding of symbols. The only reason I didn't make SymbolId public already is to give us time to pick the best encoding.

Also, SymbolKey is not public.  So we still have plenty of time to get this right.   And if and when we want to make this public we can decide on an appropriate encoding.  But we should not be coming up with arbitrary encodings and parsers for a simple feature like completion lists when we have existing components that can provide this for us.

> The whole process is going to take considerable more memory and generate a lot mor GC churn.

I'm highly skeptical that this is an issue in practice.  To give some context, the *entirety* of all TypeScript calls are done with Json.  The process converts from a managed object to Json to a managed string, to a script string, to script json, to a JS object.  That's happens for operations like Completion.  

That's not to say that we shouldn't try to be more efficient.  However, that efficiency can be an implementation detail.  Json.net makes it *trivial* to override reading and writing.  The benefit here that we're not needing to maintain our own parser.  We don't have to be concerned about any issues with encoding and whatnot, and we still use a *single* representation that is adequate for *all* our feature needs, not just the needs of completion.  

We should start with an appropriate solution first, then optimize if there are actual problems.  We should not produce a different solution which then duplicates our maintenance cost and likely means we'll never get rid of the one of them.

To be clear: i'm *extremely* concerned about the latter issue.  We've already done that with things like SyntaxGenerator and CodeGenerator.  We're building multiple overlapping systems without unification.  We then have released parts of these APIs publicly and have found ourselves unable to then change issues there that block use elsewhere because they're public.  

Finally, if we discover that json is absolutely unacceptable for some reason, then we can always switch to something handwritten.  But that something would sit on top of SymbolKey and not be an entirely new duplicated subsystem.  I'm skeptical this would ever be necessary (given that we've built entire *out of proc* systems that are using json to transfer data), but it's certainly an available option.

Note that if we do go with our own approach, that now means having to document it and write parsers for it wherever we need to use it.  Do we really want that when we already have existing interchange formats?  It seems like we're going way overboard on solving a hypothetical problem versus a cleaner design that we can optimize later freely if data indicates we need to.

**Comment by CyrusNajmabadi** [comment]
Other important points:

1. For our use cases, reading is actually not that important.  The symbols are stored in completion items, and are only read when necessary.  This only happens for certain operations like GetDescription, or when we work with a particular item.
2. Writing is important (since we may generate for many thousands of symbols), but that can be heavily optimized with very little code.  For example, we can just switch to manually using JsonTextWriter, which would be akin to the approach used by SymbolId where we write into a builder and then return that builder's contents.  If we're concerned about the SymbolKey allocations themselves, then we can pretty easily alleviate that with a little cache.  This would also be trivial to add.

In other words, SymbolKey is extremely robust and well tested, and seems exceedingly simple to make work for the needs of completion and other simple features.

**Comment by CyrusNajmabadi** [comment]
@mattwar 

I have created two alternative approaches to removing SymbolId:
Option2 - https://github.com/dotnet/roslyn/pull/11639
Option3 - https://github.com/dotnet/roslyn/pull/11645

In both of these alternate PRs I take a similar approach to yours where we encode the information directly into a stream and then we stream over that string to produce a result.  

Both of these PRs allow us to have one single SymbolKey implementation that fits the needs of *all* our consumption scenarios.  i.e. i've moved all our code over to using the single SymbolKey implementation.

The SymbolKeys support:
1. All symbols.  There are no restrictions on which symbols you can get a SymbolKey for.
2. Serialization to/from strings.
3. No in-memory object model.  i.e. the string is the representation.
4. Resolving the key in a compilation (both with and without caring about assembly identity).
5. Checking keys for equality (both with/without caring about assembly identity, and both with/without case sensitive equality checks).
6. Getting hashes for the keys.  again both with/without caring about assembly identity, and both with/without case sensitive equality checks

The difference between Option1 and Option2 is how '6' is achieved.  Option1 uses no allocations to compute hashes for symbolkeys (when assemblies are ignored).  Instead it uses more CPU.  Option2, will allocate when getting a hash (when assemblies are ignored) but it has a simpler implementation.

--

Note: as an *implementation detail* i picked S-expressions as my encoding strategy.  That was because they are probably the simplest compact format to encode/decode hierarchical data into.  However, i have structured things such that different representations can be used for every symbol type we support.

**Comment by CyrusNajmabadi** [comment]
We can now decide how we want to move forward.  If we truly believe that symbolkey object-model allocations are a problem, then Option2 or Option3 allows us to avoid them.  Either of them will suffice with that goal.  

At the same time, all options demonstrate their suitability for being a uniform solution for all our Symbol-Reference needs.

**Comment by mattwar** [comment]
I think it is possible to generate the hash without allocations (exception for maybe pools).  You can also cache the hashcode once its computed so it doesn't need to be computed again.

**Comment by CyrusNajmabadi** [comment]
Option3 generates without the allocations.  Note: i don't think the allocations for hasing are even interesting.  The scenarios where we even hash never work with more than a handful for SymbolKeys.  That's why i prefer option3 myself.  It's streamlined for all the large cases, and it's nice and simple for all the narrow cases.

**Comment by mattwar** [comment]
I'm not sure why all the old subtypes are structs in option 2, since they are not holding data. Static classes?  SymbolKey as a struct would be nice.

**Comment by mattwar** [comment]
The only downside is the serialization is not eyeball readable.  That's okay for a non-public format, but still harder to figure out what is going on while debugging.

**Comment by CyrusNajmabadi** [comment]
> I'm not sure why all the old subtypes are structs in option 2

For namespacing.  It allows us a uniform place to keep logic related to a specific symbol type.  That way they're not all cluttered together.  They should be changed to static classes for this purpose though.

**Comment by CyrusNajmabadi** [comment]
> That's okay for a non-public format, but still harder to figure out what is going on while debugging.

We could certainly embed a human readable string during debug scenarios as well.  Symbol.ToString() would be fine.
